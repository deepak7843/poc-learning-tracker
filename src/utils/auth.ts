// import { jwtDecode } from 'jwt-decode';

interface JWTPayload {
  id: string;
  email: string;
  role: string;
  exp: number;
}

const TOKEN_KEY = 'auth_token';
const REFRESH_TOKEN_KEY = 'refresh_token';

// Short-lived access token (15 minutes)
export const generateToken = (payload: Omit<JWTPayload, 'exp'>) => {
  const now = Math.floor(Date.now() / 1000);
  const exp = now + 15 * 60; // 15 minutes
  const tokenPayload = { ...payload, exp };
  return btoa(JSON.stringify(tokenPayload)); //(Base64 encode)
};

// Long-lived refresh token (7 days)
export const generateRefreshToken = (userId: string) => {
  const now = Math.floor(Date.now() / 1000);
  const exp = now + 7 * 24 * 60 * 60; // 7 days
  const payload = { userId, exp };
  return btoa(JSON.stringify(payload));
};

export const setTokens = (token: string, refreshToken: string) => {
  localStorage.setItem(TOKEN_KEY, token);
  localStorage.setItem(REFRESH_TOKEN_KEY, refreshToken);
};

export const getTokens = () => {
  return {
    token: localStorage.getItem(TOKEN_KEY),
    refreshToken: localStorage.getItem(REFRESH_TOKEN_KEY),
  };
};

export const removeTokens = () => {
  localStorage.removeItem(TOKEN_KEY);
  localStorage.removeItem(REFRESH_TOKEN_KEY);
};

export const isTokenValid = (token: string): boolean => {
  try {
    const decoded = JSON.parse(atob(token));
    const now = Math.floor(Date.now() / 1000);
    return decoded.exp > now;
  } catch {
    return false;
  }
};

export const refreshAuthToken = (userId: string, email: string, role: string) => {
  const token = generateToken({ id: userId, email, role });
  const refreshToken = generateRefreshToken(userId);
  setTokens(token, refreshToken);
  return { token, refreshToken };
};