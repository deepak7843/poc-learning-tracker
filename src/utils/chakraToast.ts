import { useToast as useChakraToast, UseToastOptions } from '@chakra-ui/react';

const defaultOptions: UseToastOptions = {
  duration: 3000,
  isClosable: true,
  position: 'top',
};


export const useToast = () => {
  const chakraToast = useChakraToast();
  
  const toast = (options: UseToastOptions): void => {
    chakraToast(options);
  };
  
  toast.success = (title: string, options?: UseToastOptions): void => {
    chakraToast({
      title,
      status: 'success',
      ...defaultOptions,
      ...options
    });
  };
  
  toast.error = (title: string, options?: UseToastOptions): void => {
    chakraToast({
      title,
      status: 'error',
      ...defaultOptions,
      ...options
    });
  };
  
  toast.warning = (title: string, options?: UseToastOptions): void => {
    chakraToast({
      title,
      status: 'warning',
      ...defaultOptions,
      ...options
    });
  };
  
  toast.info = (title: string, options?: UseToastOptions): void => {
    chakraToast({
      title,
      status: 'info',
      ...defaultOptions,
      ...options
    });
  };
  
  toast.close = (id: string): void => {
    chakraToast.close(id);
  };
  
  return toast;
};

