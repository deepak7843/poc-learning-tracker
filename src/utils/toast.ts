import { toast, type ToastOptions } from 'react-toastify';

type ToastType = 'success' | 'error' | 'info' | 'warning';
type ToastPosition = 'top-right' | 'top-center' | 'top-left' | 'bottom-right' | 'bottom-center' | 'bottom-left';

// I have used Omit**, the 'type'** property since I am handling it separately(ToastType**)
export interface ToastOptionsExtended extends Omit<ToastOptions, 'type'> {
  position?: ToastPosition;
  duration?: number;
}

const showToast = (
  message: string,
  type: ToastType = 'info',
  options: ToastOptionsExtended = {}
) => {
  const { duration = 3000, position = 'top-right', ...restOptions } = options;

  const toastOptions: ToastOptions = {
    position,
    autoClose: duration,
    hideProgressBar: false,
    closeOnClick: true,
    pauseOnHover: true,
    draggable: true,
    pauseOnFocusLoss: true,
    progress: undefined,
    theme: 'light' as const,
    ...restOptions,
  };

  switch (type) {
    case 'success':
      toast.success(message, toastOptions);
      break;
    case 'error':
      toast.error(message, toastOptions);
      break;
    case 'warning':
      toast.warn(message, toastOptions);
      break;
    case 'info':
    default:
      toast.info(message, toastOptions);
  }
};

export const toastService = {
  success: (message: string, duration?: number) =>
    showToast(message, 'success', { duration }),
  error: (message: string, duration?: number) =>
    showToast(message, 'error', { duration }),
  info: (message: string, duration?: number) =>
    showToast(message, 'info', { duration }),
  warning: (message: string, duration?: number) =>
    showToast(message, 'warning', { duration }),
};
