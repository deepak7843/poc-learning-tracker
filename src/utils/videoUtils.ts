import { VideoFile, VideoMetadata } from '../types/video';

const DB_NAME = 'VideoStorageDB';
const DB_VERSION = 1;
const STORE_VIDEOS = 'videos';
const STORE_FILES = 'videoFiles';

let dbPromise: Promise<IDBDatabase> | null = null;

const initDB = (): Promise<IDBDatabase> => {
  return new Promise((resolve, reject) => {
    if (typeof window === 'undefined' || !window.indexedDB) {
      return reject(new Error('IndexedDB is not supported'));
    }

    const request = indexedDB.open(DB_NAME, DB_VERSION);

    request.onerror = () => reject(new Error(`Failed to open database: ${request.error?.message}`));
    request.onsuccess = () => resolve(request.result);
    request.onupgradeneeded = (event) => {
      const db = (event.target as IDBOpenDBRequest).result;
      if (!db.objectStoreNames.contains(STORE_VIDEOS)) {
        db.createObjectStore(STORE_VIDEOS, { keyPath: 'id' });
      }
      if (!db.objectStoreNames.contains(STORE_FILES)) {
        // We store video data separately with the same ID
        db.createObjectStore(STORE_FILES, { keyPath: 'id' });
      }
    };
  });
};

const getDB = (): Promise<IDBDatabase> => {
  if (!dbPromise) {
    dbPromise = initDB();
  }
  return dbPromise;
};

// Helper to promisify a request, which is the standard way to work with IndexedDB
const promisifyRequest = <T>(request: IDBRequest<T>): Promise<T> => {
  return new Promise<T>((resolve, reject) => {
    request.onsuccess = () => resolve(request.result);
    request.onerror = () => reject(request.error);
  });
};

export const addVideo = async (
  file: File,
  metadata: Omit<VideoMetadata, 'id' | 'url' | 'uploadDate'>
): Promise<VideoFile> => {
  const db = await getDB();
  const videoId = `video-${Date.now()}-${file.name}`;
  const newVideo: VideoFile = {
    id: videoId,
    ...metadata,
    uploadDate: new Date().toISOString(),
    url: '',
  };

  // IMPORTANT: Read the file into memory BEFORE starting the transaction.
  // This prevents the transaction from auto-committing, which was the root cause of the error.
  const fileBuffer = await file.arrayBuffer();

  const tx = db.transaction([STORE_VIDEOS, STORE_FILES], 'readwrite');
  const videosStore = tx.objectStore(STORE_VIDEOS);
  const filesStore = tx.objectStore(STORE_FILES);

  // Wait for both 'add' operations to complete within the same transaction.
  await Promise.all([
    promisifyRequest(videosStore.add(newVideo)),
    promisifyRequest(filesStore.add({ id: videoId, data: fileBuffer })),
  ]);

  return newVideo;
};

export const readVideos = async (): Promise<VideoFile[]> => {
  const db = await getDB();
  const tx = db.transaction(STORE_VIDEOS, 'readonly');
  const store = tx.objectStore(STORE_VIDEOS);
  const videos = await promisifyRequest(store.getAll());
  return videos || [];
};

export const getVideoData = async (videoId: string): Promise<ArrayBuffer | null> => {
  const db = await getDB();
  const tx = db.transaction(STORE_FILES, 'readonly');
  const store = tx.objectStore(STORE_FILES);
  const fileObject = await promisifyRequest(store.get(videoId));
  return fileObject ? (fileObject as any).data : null;
};

export const removeVideo = async (videoId: string): Promise<void> => {
  const db = await getDB();
  const tx = db.transaction([STORE_VIDEOS, STORE_FILES], 'readwrite');
  const videosStore = tx.objectStore(STORE_VIDEOS);
  const filesStore = tx.objectStore(STORE_FILES);

  await Promise.all([
    promisifyRequest(videosStore.delete(videoId)),
    promisifyRequest(filesStore.delete(videoId)),
  ]);
};