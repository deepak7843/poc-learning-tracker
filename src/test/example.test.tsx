import { describe, it, expect } from 'vitest';
import { render, screen } from './utils';
import { ReactElement } from 'react';
import '@testing-library/jest-dom';

describe('Example Test', () => {
  it('renders without crashing', () => {
    const TestComponent = (): ReactElement => <div>Test</div>;
    render(<TestComponent />);
    const element = screen.getByText('Test');
    expect(element).toBeInTheDocument();
  });

  it('matches snapshot', () => {
    const TestComponent = (): ReactElement => <div>Test Snapshot</div>;
    const { container } = render(<TestComponent />);
    expect(container).toMatchSnapshot();
  });
});
