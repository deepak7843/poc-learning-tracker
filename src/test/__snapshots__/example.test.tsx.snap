// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Example Test > matches snapshot 1`] = `
<div>
  <div>
    Test Snapshot
  </div>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;
